### ISSUE TITLE
RulesEngine lifetime and thread-safety

### ISSUE THREAD

**Issue by Makciek** [microsoft/RulesEngine#580]
Hello
I can't find it in docs nor in wiki, what's the advised lifetime of the RulesEngine object? Should I create it every time I need it and it will cache the compiled evaluator itself or should I create just one instance and cache it? If so is it thread-safe and I can reuse the same instance at the same time?

**Comment by abbasc52** [comment]
@Makciek , it is recommended to use it as single instance to benefit from caching of compilation.

Also, it is thread safe, so you can use same instance at same time

**Comment by vanwx** [comment]
Hello, I'm curious about the thread safety of the localParam. If we aim to record the result of rule evaluation and assign it to a local parameter, could it be overwritten by other rule evaluations with different inputs?

For instance, let's say we construct a rule containing blacklist words. If the text contains any of these blacklist words, we wish to record the offensive word in the localParam and utilize it to generate the FailureMessage.

**Comment by abbasc52** [comment]
@vanwx localParam is like a local variable in code, they do not share reference with any other rule or execution. So it should be completely thread safe

**Comment by vanwx** [comment]
> @vanwx localParam is like a local variable in code, they do not share reference with any other rule or execution. So it should be completely thread safe

Yeah thanks @abbasc52 . I did write some tests and it's been safe. Also I found some examples on this repo use the localParam to store result state so it's good.
